docs/persistency.md

```diff
@@ -160,3 +160,117 @@ This is equivalent to the following command executed inside the container:
+
+3. Now it's time to prepare local/development database for development of models.py and creation of migration script.
+
+    - In `docker-compose.yml` edit database service to use volume with different name, this will create new volume and old one won't get deleted. Run `scripts/start_with_docker.sh`. This will create database that for sure reflects what's in models.py. Then to mark database with version from migrations/versions, run `flask --app yaptide.application db stamp head`. This will save id of newest version of database in alembic_version table. **Be cautios as this option is only for development on local machine.**
+
+4. Do modifications in `models.py`.
+5. Run `flask --app yaptide.application db migrate`.
+6. There will be generated migration file in migrations/versions. Name of the newest file is displayed in output of above command.
+7. **IMPORTANT!** Check the file carefully. For example there might be some `None` values which needs to be changed.
+    -  script that adds CASCADE option to foreign constraint at first looks like this
+        ```
+        def upgrade():
+            # ### commands auto generated by Alembic - please adjust! ###
+            with op.batch_alter_table('Task', schema=None) as batch_op:
+                batch_op.drop_constraint('Task_simulation_id_fkey', type_='foreignkey')
+                batch_op.create_foreign_key(None, 'Simulation', ['simulation_id'], ['id'], ondelete='CASCADE')
+        ```
+        In this case change None to 'Task_simulation_id_fkey'.
+8. Run `flask --app yaptide.application db upgrade` to apply migration script.
+9. To undo changes and go back to previous version run `flask --app yaptide.application db downgrade`.
+10. Commit and push script and modification to models.py.
+
+### Testing migration script with copy of production volume
+
+1. If there is testing environment other than local - pull changes.
+2. Copy volume ("data") of production postgres database and save it under different name.
+3. In `docker-compose.yml` modify database configuration to use this volume.
+4. Run `scripts/start_with_docker_develop.sh`, to run backend and additionaly pgadminer tool (see section Using pgadminer).
+5. Again prepare `FLASK_SQLALCHEMY_DATABASE_URI` like above and use it together with each `flask db` command.
+6. Run `flask --app yaptide.application db upgrade`.
+7. Check also `flask --app yaptide.application db downgrade`, then do upgrade again. All should execute without errors.
+8. Do manual testing. Check functionalities, some might be unnecessary to check depending which part in `models.py` was changed:
+    - logging in and out
+    - Loading simulation results, input files, logs.
+    - Submiting new simulation
+    - operations contained in `admin/db_manage.py`
+
+### Migrating production
+1. Run git pull on master.
+2. Backup the production database
+  Before applying any migrations, create a backup of the live database:
```

Comment:
You can also do full dump of the volume:

```
docker run --rm -v yaptide_data:/var/lib/postgresql/data -v /home/ubuntu/backup:/backup busybox tar czf /backup/yaptide_data_backup.tar.gz -C /var/lib/postgresql/data .
```

Reply:
I asked chatgpt about that and mentioned about risk of data inconsistency but I guess we can add that 2nd option as alternative.